File: projectAngular/src/app/services/subsidiary-bank.service.ts

import { Injectable } from '@angular/core';
import { Apollo } from 'apollo-angular';
import { gql } from 'apollo-angular';
import { Observable, map } from 'rxjs';
import {
  SubsidiaryBankDto,
  SubsidiaryBankSaveInput,
  SubsidiaryBankUpdateInput
} from '../entities/subsidiary-bank.model';

@Injectable({ providedIn: 'root' })
export class SubsidiaryBankService {
  constructor(private apollo: Apollo) { }

  getSubsidiaryBanksByBankId(bankId: string): Observable<SubsidiaryBankDto[]> {
    return this.apollo
      .watchQuery<{ getSubsidiaryBanksByBankId: SubsidiaryBankDto[] }>({
        query: gql`
          query ($bankId: ID!) {
            getSubsidiaryBanksByBankId(bankId: $bankId) {
              id
              codeSubsidiary
              name
              location
            }
          }
        `,
        variables: { bankId },
        fetchPolicy: 'cache-and-network',
      })
      .valueChanges.pipe(map(res => res.data.getSubsidiaryBanksByBankId));
  }

  saveSubsidiaryBank(bankId: string, saveDTO: SubsidiaryBankSaveInput): Observable<SubsidiaryBankDto> {
    return this.apollo.mutate<{ addSubsidiaryBank: SubsidiaryBankDto }>({
      mutation: gql`
        mutation ($bankId: ID!, $saveDTO: SubsidiaryBankSaveDTO!) {
          addSubsidiaryBank(bankId: $bankId, saveDTO: $saveDTO) {
            id
            codeSubsidiary
            name
            location
          }
        }
      `,
      variables: { bankId, saveDTO },
      refetchQueries: ['getSubsidiaryBanksByBankId'],
      awaitRefetchQueries: true
    }).pipe(map(res => res.data!.addSubsidiaryBank));
  }

  updateSubsidiaryBank(subsidiaryBankId: string, updateDTO: SubsidiaryBankUpdateInput): Observable<boolean> {
    return this.apollo.mutate<{ updateSubsidiaryBank: boolean }>({
      mutation: gql`
        mutation ($subsidiaryBankId: ID!, $updateDTO: SubsidiaryBankUpdateDTO!) {
          updateSubsidiaryBank(subsidiaryBankId: $subsidiaryBankId, updateDTO: $updateDTO)
        }
      `,
      variables: { subsidiaryBankId, updateDTO },
      refetchQueries: ['getSubsidiaryBanksByBankId'],
      awaitRefetchQueries: true
    }).pipe(map(res => res.data!.updateSubsidiaryBank));
  }

  deleteSubsidiaryBank(subsidiaryBankId: string): Observable<boolean> {
    return this.apollo.mutate<{ deleteSubsidiaryBank: boolean }>({
      mutation: gql`
        mutation ($subsidiaryBankId: ID!) {
          deleteSubsidiaryBank(subsidiaryBankId: $subsidiaryBankId)
        }
      `,
      variables: { subsidiaryBankId },
      refetchQueries: ['getSubsidiaryBanksByBankId'],
      awaitRefetchQueries: true
    }).pipe(map(res => res.data!.deleteSubsidiaryBank));
  }
}

File: BackEnd/src/main/java/com/example/backend/services/SubsidiaryBankService.java
package com.example.backend.services;

import com.example.backend.dtos.SubsidiaryBankDto;
import com.example.backend.dtos.SubsidiaryBankSaveDTO;
import com.example.backend.dtos.SubsidiaryBankUpdateDTO;
import com.example.backend.exceptions.AlreadyExistsException;
import com.example.backend.exceptions.BankNotFoundException;
import com.example.backend.exceptions.ParentEntityHasChildrenException;
import com.example.backend.exceptions.SubsidiaryBankNotFoundException;

import java.util.List;

public interface SubsidiaryBankService {

    SubsidiaryBankDto addSubsidiaryBank(Long bankId, SubsidiaryBankSaveDTO saveDTO) throws BankNotFoundException, AlreadyExistsException;

    void updateSubsidiaryBank(Long subsidiaryBankId, SubsidiaryBankUpdateDTO updateDTO) throws SubsidiaryBankNotFoundException;

    List<SubsidiaryBankDto> getSubsidiaryBanksByBankId(Long bankId);

    SubsidiaryBankDto getSubsidiaryBankById(Long subsidiaryBankId) throws SubsidiaryBankNotFoundException;

    void deleteSubsidiaryBank(Long subsidiaryBankId) throws SubsidiaryBankNotFoundException, ParentEntityHasChildrenException;
}
